packages/schemas/src/types/system.ts

```diff
@@ -64,6 +64,42 @@ export const storageProviderGuard: Readonly<{
   [StorageProviderKey.StorageProvider]: storageProviderDataGuard,
 });
 
+// Email service provider
+export enum EmailServiceProvider {
+  SendGrid = 'SendGrid',
+}
+
+export const emailServiceDataGuard = z.discriminatedUnion('provider', [
+  z.object({
+    provider: z.literal(EmailServiceProvider.SendGrid),
+    appId: z.string(),
+    appSecret: z.string(),
+    fromEmail: z.string(),
+    templates: z.object({
+      general: z.object({
+        subject: z.string(),
+        content: z.string(),
+      }),
+    }),
+  }),
+]);
+
+export type EmailServiceData = z.infer<typeof emailServiceDataGuard>;
+
+export enum EmailServiceProviderKey {
+  EmailServiceProvider = 'EmailServiceProvider',
+}
+
+export type EmailServiceProviderType = {
+  [EmailServiceProviderKey.EmailServiceProvider]: EmailServiceData;
+};
+
+export const emailServiceProviderGuard: Readonly<{
+  [key in EmailServiceProviderKey]: ZodType<EmailServiceProviderType[key]>;
+}> = Object.freeze({
+  [EmailServiceProviderKey.EmailServiceProvider]: emailServiceDataGuard,
+});
```

Comment:
Why don't we define the guard first and use `z.infer` to generate the related type?

Reply:
this implementation aligns with other system key definitions. Maybe it's because we need to specify the key in zod object should be some enum types.